### PR TITLE
Attach session compaction receipts to text requests

### PR DESCRIPTION
**Pull request by Zigfreidish**
## Summary
- Attach the existing session compaction policy receipt to accepted live text requests before worker dispatch.
- Thread resolved session/model/request budget context from the OpenAI gateway into `ChatRequestTranslator` without mutating stored session history.
- Add focused translator and gateway tests covering normal, over-budget, missing-context, media, empty-part, and invalid setting paths.
- Address review feedback by replacing `split(whereSeparator:)` token word counting with a single-pass state machine.

## Plan or Spec
- Refs #1383.
- Governing plan: `docs/plans/2026-07-09-issue-1383-session-compaction-policy.md`.

## Commands Run
```text
make bootstrap
# Passed: configured .githooks and confirmed services/mlx-worker-python dependencies from the locked environment.

make proto
# Passed: ./scripts/proto_gen.sh completed; no generated diff.

HOME="$PWD/.swift-home" CLANG_MODULE_CACHE_PATH="$PWD/.build/ModuleCache.noindex" xcrun swift test --package-path services/control-plane-swift --filter 'ControlPlaneTests.TextEndpointContractTests/sessionCompactionPolicy'
# Baseline passed: 6 tests.

HOME="$PWD/.swift-home" CLANG_MODULE_CACHE_PATH="$PWD/.build/ModuleCache.noindex" xcrun swift test --package-path services/control-plane-swift --filter 'chatTranslationAttachesSessionCompactionReceiptBeforeWorkerDispatch|chatTranslationMarksCompactionRequiredWhenRetainedSessionItemsExceedUsableContext|gatewaySessionCompactionReceiptFeedsAcceptedTextRequests'
# TDD red first failed to compile for the missing request context and translator argument, then passed after implementation: 3 tests.

HOME="$PWD/.swift-home" CLANG_MODULE_CACHE_PATH="$PWD/.build/ModuleCache.noindex" xcrun swift test --package-path services/control-plane-swift --filter 'ControlPlaneTests.TextEndpointContractTests/sessionCompactionPolicy'
# Passed: 6 tests.

HOME="$PWD/.swift-home" CLANG_MODULE_CACHE_PATH="$PWD/.build/ModuleCache.noindex" xcrun swift test --package-path services/control-plane-swift --filter 'ControlPlaneTests.TextEndpointContractTests'
# Passed: 82 tests.

HOME="$PWD/.swift-home" CLANG_MODULE_CACHE_PATH="$PWD/.build/ModuleCache.noindex" xcrun swift test --package-path services/control-plane-swift --filter 'promptBudget|SessionCompaction'
# Passed: 9 focused OpenAI gateway/session compaction tests.

HOME="$PWD/.swift-home" CLANG_MODULE_CACHE_PATH="$PWD/.build/ModuleCache.noindex" xcrun swift test --package-path services/control-plane-swift --filter 'ControlPlaneTests.TextEndpointContractTests|promptBudget|SessionCompaction'
# Passed: 91 tests.

HOME="$PWD/.swift-home" CLANG_MODULE_CACHE_PATH="$PWD/.build/ModuleCache.noindex" xcrun swift test --package-path services/control-plane-swift --filter 'ControlPlaneTests.TextEndpointContractTests/sessionCompactionPolicy|chatTranslationAttachesSessionCompactionReceiptBeforeWorkerDispatch|chatTranslationMarksCompactionRequiredWhenRetainedSessionItemsExceedUsableContext|gatewaySessionCompactionReceiptFeedsAcceptedTextRequests|chatTranslationEstimatesMediaAndEmptyPartsInSessionCompactionReceipts'
# Review follow-up passed: 10 tests.

HOME="$PWD/.swift-home" CLANG_MODULE_CACHE_PATH="$PWD/.build/ModuleCache.noindex" xcrun swift test --package-path services/control-plane-swift --enable-code-coverage --filter 'ControlPlaneTests.TextEndpointContractTests/sessionCompactionPolicy|chatTranslationAttachesSessionCompactionReceiptBeforeWorkerDispatch|chatTranslationMarksCompactionRequiredWhenRetainedSessionItemsExceedUsableContext|chatTranslationSkipsSessionCompactionReceiptWithoutResolvedSessionContext|chatTranslationEstimatesMediaAndEmptyPartsInSessionCompactionReceipts|gatewaySessionCompactionReceiptFeedsAcceptedTextRequests|gatewaySkipsSessionCompactionReceiptWhenModelContextIsUnknown|gatewayDefaultsInvalidSessionCompactionHistorySettingsToUnlimited'
# Passed: 13 tests with Swift coverage enabled.

python3 scripts/swift_changed_line_coverage.py --binary ...
# Failed before execution because system python is too old for the script's type syntax.

UV_PYTHON=3.12 uv run python scripts/swift_changed_line_coverage.py --binary services/control-plane-swift/.build/arm64-apple-macosx/debug/MelixControlPlanePackageTests.xctest/Contents/MacOS/MelixControlPlanePackageTests --profdata services/control-plane-swift/.build/arm64-apple-macosx/debug/codecov/default.profdata --diff-from origin/main services/control-plane-swift/Sources/Requests/SessionCompactionPolicy.swift services/control-plane-swift/Sources/Requests/ChatRequestTranslator.swift services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift services/control-plane-swift/Tests/ControlPlaneTests/TextEndpointContractTests.swift services/control-plane-swift/Tests/HTTPGatewayTests/OpenAIHandlerTests.swift
# Passed: changed-line coverage TOTAL 99.39%, 491/494.

git merge origin/main
# Passed: merged origin/main 11b328a7 with no conflicts.

HOME="$PWD/.swift-home" CLANG_MODULE_CACHE_PATH="$PWD/.build/ModuleCache.noindex" xcrun swift test --package-path services/control-plane-swift --filter 'ControlPlaneTests.TextEndpointContractTests/sessionCompactionPolicy|chatTranslationAttachesSessionCompactionReceiptBeforeWorkerDispatch|chatTranslationMarksCompactionRequiredWhenRetainedSessionItemsExceedUsableContext|chatTranslationSkipsSessionCompactionReceiptWithoutResolvedSessionContext|chatTranslationEstimatesMediaAndEmptyPartsInSessionCompactionReceipts|gatewaySessionCompactionReceiptFeedsAcceptedTextRequests|gatewaySkipsSessionCompactionReceiptWhenModelContextIsUnknown|gatewayDefaultsInvalidSessionCompactionHistorySettingsToUnlimited'
# Post-merge passed: 13 tests.

git diff --check
# Passed.

make swift-test
# Passed. Final stage: menubar suite 834 tests passed.

make py-test
# Passed: 4908 passed, 14 skipped, 2 warnings.

make integration-test
# Passed: 123 passed, 1 skipped.

.githooks/pre-commit
# Passed manually. Full local gate passed and wrote performance report .runtime/pre-commit-performance/20260711-151106-7ec7e64c/report/report.md.

git commit
# Passed pre-commit hook before initial commit 48686e85. Hook report path: .runtime/pre-commit-performance/20260711-153834-7ec7e64c/report/report.md.

git commit --amend --no-edit
# Passed pre-commit hook after review fix before amended commit 2bb2b8de. Hook results: make swift-test passed; make py-test passed (4908 passed, 14 skipped, 2 warnings); make integration-test passed (123 passed, 1 skipped); performance report status ok at .runtime/pre-commit-performance/20260711-160603-48686e85/report/report.md.
```

## Coverage and Metrics
- Changed-line coverage after review follow-up: 99.39% total, 491/494 covered.
- `SessionCompactionPolicy.swift`: 98.51%, 66/67; uncovered line 82.
- `ChatRequestTranslator.swift`: 100.00%, 9/9.
- `OpenAIHandler.swift`: 96.30%, 52/54; uncovered lines 4053 and 4091.
- `TextEndpointContractTests.swift`: 100.00%, 192/192.
- `OpenAIHandlerTests.swift`: 100.00%, 172/172.
- Latest local pre-commit performance report: `.runtime/pre-commit-performance/20260711-160603-48686e85/report/report.md`; status ok, changed files 1, selected probes 0, regressions 0.
- PR scoped performance workflow report: status ok, changed files 6, selected probes 0, regressions 0.
- Observability mode: `off`; this slice reuses request extension receipts and does not add runtime metrics, evidence probes, or debug diagnostics.
- Probe overhead: `N/A`; no new probes were added.

## Known Gaps
- This is a planner-only live request assembly slice. It attaches compaction receipts to worker request extensions but does not trim, summarize, or mutate stored session history.
- Non-blocking test gap: two defensive gateway branches remain uncovered (`no session_id` early return and empty `model.modelID` fallback), with changed-line coverage still at 99.39% for the touched Swift scope.

## Evidence Checklist
- [x] Relevant plan or spec is identified.
- [x] Behavior changes are reflected in the relevant docs.
- [x] Protocol changes include regenerated generated artifacts. `N/A`: no protocol schema changes; `make proto` was run and produced no diff.
- [x] Dependency changes include updated lockfiles. `N/A`: no dependency changes.
- [x] Relevant tests were run.
- [x] A metrics report is included, or `N/A` is stated explicitly with the reason.
- [x] Observability mode, probe overhead, and any deferred debug or evidence work are recorded, or `N/A` is stated explicitly with the reason.
- [x] Deferred work and known gaps are stated explicitly.

### CODE REVIEW

File: services/control-plane-swift/Sources/Requests/SessionCompactionPolicy.swift [hunk 1]
```diff
@@ -1,4 +1,90 @@
...
+
+    private static func tokenCount(in text: String) -> UInt32 {
+        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard !trimmed.isEmpty else {
+            return 0
+        }
+        let whitespaceEstimate = trimmed.split(whereSeparator: \.isWhitespace).count
+        let byteEstimate = max(1, (trimmed.utf8.count + 3) / 4)
+        return UInt32(min(Int(UInt32.max), max(whitespaceEstimate, byteEstimate)))
+    }
```
[the remark this answers is not in the document]
Comment (Zigfreidish) [reply services/control-plane-swift/Sources/Requests/SessionCompactionPolicy.swift:90]: Fixed in 2bb2b8de. `tokenCount(in:)` now counts word transitions with a single-pass state machine instead of allocating `Substring` arrays through `split(whereSeparator:)`, while preserving the trimmed text and byte-estimate semantics. Verification: focused review follow-up Swift filter passed 10 tests; coverage-enabled focused Swift filter passed 13 tests; changed-line coverage is 99.39% (491/494); amended pre-commit passed `make swift-test`, `make py-test`, `make integration-test`, and the scoped performance report with 0 regressions.

File: services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift [hunk 2]
```diff
@@ -4035,6 +4041,56 @@ button.primary:active {
...
+    private func sessionCompactionRequestContext(
+        normalized: NormalizedTextRequest,
+        model: Melix_Controlplane_V1_ModelSummary?,
+        modelSamplingPolicy: ModelSamplingPolicy?,
+        gatewayServingDefaults: GatewayServingDefaultsPolicy?
+    ) -> SessionCompactionRequestContext? {
+        guard let sessionID = normalized.sessionID?.trimmingCharacters(in: .whitespacesAndNewlines),
+              !sessionID.isEmpty
+        else {
+            return nil
```
Comment (Zigfreidish) [inline services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift:4053]: Minor, non-blocking: this is the one gateway-level branch not exercised by the new tests — every `gatewaySessionCompactionReceipt*` test sends a `session_id`. Matches the PR's own coverage report (line 4053 listed as uncovered). Worth a quick follow-up test for the "no session_id at all" case, but the logic itself mirrors the `contextWindowTokens > 0` check right below it and looks correct.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift [hunk 3]
```diff
@@ -4035,6 +4041,56 @@ button.primary:active {
...
+
+    private func normalizedTextModelID(
+        normalized: NormalizedTextRequest,
+        model: Melix_Controlplane_V1_ModelSummary?
+    ) -> String {
+        let modelID = model?.modelID.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
+        if !modelID.isEmpty {
+            return modelID
+        }
+        return normalized.model.trimmingCharacters(in: .whitespacesAndNewlines)
```
Comment (Zigfreidish) [inline services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift:4091]: Similarly uncovered (PR's report lists line 4091): the `model?.modelID` empty fallback to `normalized.model` never fires in the new tests since `warmModel()` always has a `modelID`. Not a bug — just flagging so it doesn't silently rot — the fallback path itself reads correctly.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by Zigfreidish — COMMENTED** [review]
Reviewed the diff manually. The implementation matches the plan doc (`docs/plans/2026-07-09-issue-1383-session-compaction-policy.md`): planner-only, no session-history mutation, and the receipt attachment is threaded correctly from `OpenAIHandler` → `ChatRequestTranslator` → `SessionCompactionPolicy.plan`.

Verified by hand:
- `tokenCount(in:)`'s single-pass rewrite (replacing `split(whereSeparator:)`) produces the same word/byte-based estimates the tests expect — traced the arithmetic for the plain-text, dense-text, and media/empty-part test cases and all match.
- Guard ordering in `sessionCompactionRequestContext` (blank session → nil, zero context window → nil, `usableContextTokens` underflow avoided via the `contextWindowTokens > outputCapTokens` check) is sound.
- `parsePositiveUInt32` correctly defaults invalid/zero `max_history_items` settings to "unlimited" as documented.
- Merge into `execution.ext` uses distinct `melix.session_compaction.*` keys, no collision risk with existing `openAICompatibilityReceipts` merge.

Left two non-blocking inline notes on gateway branches (no `session_id` at all; empty `model.modelID` fallback) that are logically fine but aren't exercised by the new tests, per the PR's own coverage report (52/54 lines). No correctness issues found — this would be an approval, but GitHub blocks self-approval since this PR and my review both run under the same account (`Zigfreidish`). Recommend merging once a human/other account can formally approve, or if branch protection doesn't require it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift (hunk 2 above)
Comment (Zigfreidish) [reply services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift:4053]: No code change for this thread. This branch is the defensive early return for requests without `session_id`, so it correctly skips receipt assembly before any model-budget work. The behavior is not a correctness gap for this slice, and touched Swift changed-line coverage remains 99.39% overall. I recorded it as a non-blocking test gap in the PR evidence rather than expanding scope here.

File: services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift (hunk 3 above)
Comment (Zigfreidish) [reply services/control-plane-swift/Sources/HTTPGateway/OpenAI/OpenAIHandler.swift:4091]: No code change for this thread. The empty `model.modelID` path is a defensive fallback to the normalized request model for malformed or incomplete model summaries; normal gateway fixtures use non-empty model IDs, and the fallback does not alter the live receipt contract. I recorded it as a non-blocking test gap in the PR evidence; changed-line coverage remains above the repository threshold at 99.39%.
